kinesis-client/src/main/java/com/accenture/rig/RecordProcessor.java

```diff
@@ -0,0 +1,95 @@
+    this.erlangInterface = erlangInterface;
+  }
+
+  /**
+   * Invoked by the Amazon Kinesis Client Library before data records are delivered to the RecordProcessor instance
+   * (via processRecords).
+   *
+   * @param initializationInput Provides information related to initialization
+   */
+  @Override
+  public void initialize(final InitializationInput initializationInput) {
+    shardId = initializationInput.getShardId();
+    LOG.info("Initializing record processor for shard: " + shardId);
+  }
+
+  /**
+   * Process data records. The Amazon Kinesis Client Library will invoke this method to deliver data records to the
+   * application.
+   * Upon fail over, the new instance will get records with sequence number > checkpoint position
+   * for each partition key.
+   *
+   * @param processRecordsInput Provides the records to be processed as well as information and capabilities related
+   *                            to them (eg checkpointing).
+   */
+  @Override
+  public void processRecords(final ProcessRecordsInput processRecordsInput) {
+    final List<Record> records = processRecordsInput.getRecords();
+    LOG.info(String.format("Shard %s: received %d records for processing", shardId, records.size()));
+    records.stream().map(record -> {
+      final Map<String, Object> map = new HashMap<>();
+      map.put("shardId", shardId);
+      map.put("partitionKey", record.getPartitionKey());
+      map.put("sequenceNumber", record.getSequenceNumber());
+      map.put("approximateArrivalTimestamp", record.getApproximateArrivalTimestamp());
+      map.put("body", record.getData());
+      return map;
+    }).forEach(recordMap -> {
+      try {
+        erlangInterface.forward(recordMap);
+      } catch (IOException | OtpErlangDecodeException | OtpErlangExit | OtpAuthException e) {
+        final String incoming = recordMap.entrySet().stream()
+            .map(entry -> String.format("%s => %s", entry.getKey(), entry.getValue()))
+            .collect(Collectors.joining(", "));
+        LOG.error(String.format("Failed to forward message: %s", incoming), e);
+        // Fail fast - RIG will supervise/restart...
+        System.exit(1);
+      }
+    });
+  }
+
+  /**
+   * Invoked by the Amazon Kinesis Client Library to indicate it will no longer send data records to this
+   * RecordProcessor instance.
+   * <p>
+   * <h2><b>Warning</b></h2>
+   * <p>
+   * When the value of {@link ShutdownInput#getShutdownReason()} is
+   * {@link ShutdownReason#TERMINATE} it is required that you
+   * checkpoint. Failure to do so will result in an IllegalArgumentException, and the KCL no longer making progress.
+   *
+   * @param shutdownInput Provides information and capabilities (eg checkpointing) related to shutdown of this record processor.
+   */
+  @Override
+  public void shutdown(final ShutdownInput shutdownInput) {
+    LOG.info("Shutting down record processor for shard: " + shardId);
```

Comment:
Was playing a bit with cluster and multiple shards and seems the problem is stated directly in method definition

> When the value of {@link ShutdownInput#getShutdownReason()} is {@link ShutdownReason#TERMINATE} it is required that you checkpoint. Failure to do so will result in an IllegalArgumentException, and the KCL no longer making progress.

I tried just simple code and seems that re-sharding started to work.

```
try {
      shutdownInput.getCheckpointer().checkpoint();
    } catch (Exception e) {
      //TODO: handle exception
    }
```

Ideally we should probably check for reason and if it's `terminate` then do checkpoint.

One odd thing I found is that if you do re-sharding from e.g. 1 to 2, new consumers can get original (closed) shard. Not sure if that's desirable, but I guess it is. It's just a bit weird that with 2 shards you may need to run 3 consumers so these 2 new shards are spread amongst consumers.

Reply:
ah I see, should've put more attention to the docs ^_^ thanks!